### PR TITLE
crud for department and teams

### PR DESCRIPTION
**Pull request by Sumbal-Tariq**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -14,15 +12,18 @@ gem 'uglifier', '>= 1.3.0'
...
 gem 'jquery-rails'
 # Turbolinks makes following links in your web application faster. Read more: https://github.com/rails/turbolinks
-gem 'turbolinks'
+# gem 'turbolinks'
 # Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
 gem 'jbuilder', '~> 2.0'
+# Tokeninput jQuery plugin which allows your users to select multiple items from a predefined list
+gem 'rails-jquery-tokeninput'
 # bundle exec rake doc:rails generates the API under doc/api.
 gem 'sdoc', '~> 0.4.0', group: :doc
```
Comment (tehreemtalat7) [inline Gemfile]: Add in the group :development block

File: Gemfile [hunk 2]
```diff
@@ -36,11 +37,14 @@ gem 'sdoc', '~> 0.4.0', group: :doc
 group :development, :test do
   # Call 'byebug' anywhere in the code to stop execution and get a debugger console
   gem 'byebug'
```
Comment (tehreemtalat7) [inline Gemfile]: remove byebug if using pry

File: app/models/company.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class Company < ActiveRecord::Base
+  belongs_to :employee
```
Comment (tehreemtalat7) [inline app/models/company.rb]: if company has_many employees, why belong_to employee?

File: app/models/team.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  has_many :employee_teams
+  has_many :employees, through: :employee_teams, dependent: :destroy
+  validates :name, :department_id, presence: true
+
+  def create_team(team_lead_id, employee_ids)
+    team_members_ids = employee_ids(employee_ids)
+    begin
+      self.transaction do
+        self.save!
+        self.employee_teams.create!(employee_id: team_lead_id, employee_type: 'Team Leader')
```
Comment (tehreemtalat7) [inline app/models/team.rb]: use constants for employee_type

File: app/models/team.rb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        if team_members_ids.empty?
+          self.errors.add(:base, 'Team members name required')
+          return false
+        end
+        team_members_ids.each do |team_member|
+          self.employee_teams.create!(employee_id: team_member, employee_type: 'Team Member')
+        end
+        return self
+      end
+    rescue ActiveRecord::RecordInvalid => e
```
Comment (tehreemtalat7) [inline app/models/team.rb]: rename e?

File: app/models/team.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+      if error.record.employee_id.nil? && error.record.employee_type == 'Team Leader'
+        self.errors.add(:base, 'Team leader name is required')
+      end
+      if error.record.employee_id.nil? && error.record.employee_type == 'Team Member'
+        self.errors.add(:base, 'Team member names are required')
+      end
+    end
+  end
+
+  def employee_ids(employee_ids)
```
Comment (tehreemtalat7) [inline app/models/team.rb]: rename this function. its use gets ambiguous

File: app/models/team.rb [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      false
+    end
+  end
+
+  def update_team(team_lead_id, employee_ids)
+    team_members_ids = employee_ids(employee_ids)
+    begin
+      self.transaction do
+        self.save!
+        team_employees = self.employee_teams
```
Comment (tehreemtalat7) [inline app/models/team.rb]: probably the lines where we're getting/finding from database can be moved out of the transaction?

File: app/models/company.rb (hunk 3 above)
Comment (Sumbal-Tariq) [reply app/models/company.rb]: For admin id purpose... I don't have any idea about this.. Zain, please review this again...

File: app/controllers/departments_controller.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
+class DepartmentsController < ApplicationController
+  before_action :set_department, only: [:show, :edit, :update, :destroy]
+  def index
+    @departments = Department.all
```
Comment (zainsaleem4338) [inline app/controllers/departments_controller.rb]: current_employee.company.departments.all

File: app/controllers/departments_controller.rb [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+  def new
+    @department = Department.new
+  end
+
+  def edit
+  end
+
+  def create
+    @department = Department.new(department_params)
```
Comment (zainsaleem4338) [inline app/controllers/departments_controller.rb]: current_employee.company.departments.new(department_params)

and in the edit as well

File: app/controllers/departments_controller.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        format.json { render :show, status: :ok, location: @department }
+      else
+        format.html { render :edit }
+        format.json { render json: @department.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  def destroy
+    @department.destroy
```
Comment (zainsaleem4338) [inline app/controllers/departments_controller.rb]: check failure case as well
@deparment.destroyed?

File: app/controllers/employees_controller.rb [hunk 11]
```diff
@@ -3,6 +3,13 @@ def index
     @employees = current_employee.company.employees.active_members
   end
 
+  def employees_lists
+    @employees = Employee.order(:name)
```
Comment (zainsaleem4338) [inline app/controllers/employees_controller.rb]: current_employee.company.employees.order(:name)

File: app/controllers/teams_controller.rb [hunk 12]
```diff
@@ -0,0 +1,57 @@
+class TeamsController < ApplicationController
+  before_action :set_team, only: [:show, :edit, :update, :destroy]
+  def index
+    @teams = Team.all
```
Comment (zainsaleem4338) [inline app/controllers/teams_controller.rb]: current_employee.company.teams

File: app/controllers/teams_controller.rb [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+  def new
+    @team = Team.new
+  end
+
+  def edit
+  end
+
+  def create
+    @team = Team.new(team_params)
```
Comment (zainsaleem4338) [inline app/controllers/teams_controller.rb]: same thing here

File: app/controllers/teams_controller.rb [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+      else
+        format.html { render :new }
+        format.json { render json: @team.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  def update
+    respond_to do |format|
+      if @team.update_team(params[:team][:team_lead_id], params[:team][:employee_tokens])
```
Comment (zainsaleem4338) [inline app/controllers/teams_controller.rb]: use strong parameters

File: app/controllers/teams_controller.rb [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+        format.json { render :show, status: :created, location: @team }
+      else
+        format.html { render :new }
+        format.json { render json: @team.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  def destroy
+    @team.destroy
```
Comment (zainsaleem4338) [inline app/controllers/teams_controller.rb]: destroyed?

File: app/models/team.rb [hunk 16]
```diff
@@ -0,0 +1,83 @@
+class Team < ActiveRecord::Base
+  EMPLOYEE_TYPE = ['Team Member', 'Team Leader'].freeze
```
Comment (zainsaleem4338) [inline app/models/team.rb]: make it hash
because EMPLOYEE_TYPE[1] is not very explanatory

File: app/models/team.rb [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+  has_many :employee_teams
+  has_many :employees, through: :employee_teams, dependent: :destroy
+  validates :name, :department_id, presence: true
+
+  def create_team(team_lead_id, employee_ids)
+    team_members_ids = get_employee_ids(employee_ids)
+    begin
+      self.transaction do
+        self.save!
+        self.employee_teams.create!(employee_id: team_lead_id, employee_type: EMPLOYEE_TYPE[1])
```
Comment (zainsaleem4338) [inline app/models/team.rb]: use hash

File: app/models/team.rb [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+  def update_team(team_lead_id, employee_ids)
+    team_members_ids = get_employee_ids(employee_ids)
+    begin
+      if team_members_ids.empty?
+        self.errors.add(:base, 'Team members name required')
+        return false
+      end
+      team_employees = self.employee_teams
+      team_lead = team_employees.find_by(employee_type: EMPLOYEE_TYPE[1])
+      team_members = team_employees.where(employee_type: EMPLOYEE_TYPE[0]).pluck(:employee_id)
```
Comment (zainsaleem4338) [inline app/models/team.rb]: use collect instead of pluck

tell me why I added this comment? :P

File: app/views/departments/_form.html.erb [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+      <% @department.errors.full_messages.each do |message| %>
+        <li><%= message %></li>
+      <% end %>
+      </ul>
+    </div>
+  <% end %>
+
+  <div class="form-group">
+    <%= f.label :name %><br>
+    <%= f.text_field :name, :required => true, class: "form-control" %>
```
Comment (zainsaleem4338) [inline app/views/departments/_form.html.erb]: required: true (new syntax)

File: config/database.yml [hunk 20]
```diff
@@ -1,15 +1,21 @@
-# SQLite version 3.x
-#   gem install sqlite3
+# MySQL.  Versions 5.0+ are recommended.
 #
-#   Ensure the SQLite 3 gem is defined in your Gemfile
-#   gem 'sqlite3'
+# Install the MYSQL driver
```
Comment (zainsaleem4338) [inline config/database.yml]: remove this commented out code please

**Comment by zainsaleem4338** [comment]
cycle 1

**Pull request merged**